nrf-softdevice/src/ble/advertisement_builder.rs

```diff
@@ -138,145 +160,163 @@ pub enum BasicService {
-impl Service for BasicService {
-    const SIZE: usize = 2;
-
-    fn render<const N: usize>(self, adv: &mut AdvertisementData<N>) {
-        let data = (self as u16).swap_bytes().to_be_bytes();
-        adv.write(&data);
-    }
-}
-
-pub struct CustomService(pub [u8; 16]);
-
-impl Service for CustomService {
-    const SIZE: usize = 16;
-
-    fn render<const N: usize>(mut self, adv: &mut AdvertisementData<N>) {
-        self.0.reverse();
-        adv.write(&self.0);
-    }
-}
-
-pub struct Incomplete16<const N: usize>(pub [BasicService; N]);
-pub struct Complete16<const N: usize>(pub [BasicService; N]);
-pub struct Incomplete128<const N: usize>(pub [CustomService; N]);
-pub struct Complete128<const N: usize>(pub [CustomService; N]);
-
-macro_rules! impl_service_list {
-    ($LIST:ident, $SERVICE:ident, $AD:ident) => {
-        impl<const N: usize> ServiceList<$SERVICE, N> for $LIST<N> {
-            const AD: ADType = ADType::$AD;
-
-            fn list(self) -> [$SERVICE; N] {
-                self.0
-            }
-        }
-    };
+pub struct AdvertisementBuilder<const N: usize> {
+    buf: [u8; N],
+    ptr: usize,
 }
 
-impl_service_list!(Incomplete16, BasicService, Incomplete16ServiceList);
-impl_service_list!(Complete16, BasicService, Complete16ServiceList);
-impl_service_list!(Incomplete128, CustomService, Incomplete128ServiceList);
-impl_service_list!(Complete128, CustomService, Complete128ServiceList);
-
-pub trait Name {
-    const AD: ADType;
-
-    fn inner(&self) -> &str;
+pub struct AdvertisementPayload<const N: usize> {
+    buf: [u8; N],
+    len: usize,
 }
 
-pub struct ShortName<'a>(pub &'a str);
-pub struct FullName<'a>(pub &'a str);
-
-macro_rules! impl_name {
-    ($NAME:ident, $AD:ident) => {
-        impl<'a> Name for $NAME<'a> {
-            const AD: ADType = ADType::$AD;
-
-            fn inner(&self) -> &str {
-                self.0
-            }
-        }
-    };
+impl<const N: usize> AsRef<[u8]> for AdvertisementPayload<N> {
+    fn as_ref(&self) -> &[u8] {
+        &self.buf[..self.len]
+    }
 }
 
-impl_name!(ShortName, ShortName);
-impl_name!(FullName, FullName);
+impl<const N: usize> core::ops::Deref for AdvertisementPayload<N> {
+    type Target = [u8];
 
-pub struct AdvertisementData<const N: usize> {
-    buf: [u8; N],
-    ptr: usize,
+    fn deref(&self) -> &Self::Target {
+        &self.buf[..self.len]
+    }
 }
 
-impl<const K: usize> AdvertisementData<K> {
-    pub fn new() -> Self {
+impl<const K: usize> AdvertisementBuilder<K> {
+    pub const fn new() -> Self {
         Self { buf: [0; K], ptr: 0 }
     }
 
-    fn write(&mut self, data: &[u8]) {
+    const fn write(mut self, data: &[u8]) -> Self {
         let end = self.ptr + data.len();
 
-        if end <= K {
-            self.buf[self.ptr..end].copy_from_slice(data);
+        let mut i = 0;
+        while self.ptr < K && i < data.len() {
+            self.buf[self.ptr] = data[i];
+            i += 1;
+            self.ptr += 1;
         }
```

Comment:
What is the motivation behind this? If the data does not fit, why write any of it?